### ISSUE TITLE
Figure out how to remove Java thread local Context

### ISSUE THREAD

**Issue by gmethvin** [playframework/playframework#6168]
Java `Http.Context.current()` uses a thread local. Play 2.6 should provide an easy way to migrate to something that does not use thread local state.

One solution might be to make Java method calls look like:

``` java
public Result foo(Context cxt, String param1, int param2) {
  // ...
}
```

If we did that, we'd need to decide if there should be a special syntax in the routes for adding the context.

Another idea is

``` java
public FunctionalResult foo(String param1, int param2) {
  return ctx -> {
    // ...
  }
}
```

That might be a starting point for the evolution toward a more functional API.

I see this as a more important goal than creating a completely new API for Java actions, which is something that could potentially wait for 3.0.

**Comment by marcospereira** [comment]
Not sure if this is possible, but maybe another solution is to change `play.mvc.Controller` to have a `Context` field that is injected each request not having controllers as singletons. Of course, this will possible only if Guice can use field/method injection after creating an instance via constructor injection.

**Comment by gmethvin** [comment]
@marcospereira Yeah, that could work, but like you said, it requires controllers to be singletons, and I'm not sure if that will work for everyone.

If you actually are creating controllers on each request, you could actually do it with constructor injection though.

**Comment by marcospereira** [comment]
@gmethvin I have the impression that requiring a `Context` param or returning an `FunctionalResult` could be more intrusive (and harder to upgrade) than requiring controllers to be singletons. That was also why I didn't suggested constructor injection. 

Anyway, I think we can also assume that a lot of people are using `Http.Context` out of controllers scope, which will require some sort of refactoring from their part.

**Comment by wsargent** [comment]
@marcospereira We could elide the issue to some extent by making `Context` a feature:

```
public Result foo(Context cxt, String param1, int param2) {
  // ...
}
```

and deprecating Http.Context.current() in 2.6 for controllers that don't have Context:

```
public Result foo(String param1, int param2) {
  // ...
}
```

**Comment by marcospereira** [comment]
@wsargent how would that impact routes/router?

**Comment by wsargent** [comment]
@marcospereira we'd provide two methods for every route?

**Comment by tajakobsen** [comment]
Is it possible to have a `ContextProvider` or `ContextFactory` that can be injected into `Controller` (also singletons). And use it to get a `Context` returned?

**Comment by gmethvin** [comment]
@tajakobsen I think that would still require a thread local, though. How do we know which action we are in when we call the `ContextProvider`?

**Comment by wsargent** [comment]
Also worth keeping in mind the people who want an @Inject of Http.Context in a "request scope".  Guice does have the ability to define [custom scopes](https://github.com/google/guice/wiki/CustomScopes):

```
import com.google.common.collect.Maps;
import com.google.inject.Key;
import com.google.inject.Provider;
import com.google.inject.Scope;
import com.google.inject.Scopes;
import play.mvc.Http;

import java.util.Map;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Allows objects to be bound to Play! Http.Context.current.args with a ThreadLocal fallback.
 */
public class HttpContextScope implements Scope {

    private static final ThreadLocal<Context> httpContextScopeContext = new ThreadLocal<>();

    enum NullableObject {
        INSTANCE
    }

    @Override
    public <T> Provider<T> scope(final Key<T> key, final Provider<T> provider) {
        return new Provider<T>() {
            @Override
            public T get() {
                Http.Context currentContext = Http.Context.current();
                if (currentContext == null) {
                    Context context = httpContextScopeContext.get();
                    if (context != null) {
                        T t = (T) context.map.get(key);
                        if (t == NullableObject.INSTANCE) {
                            return null;
                        }

                        if (t == null) {
                            t = provider.get();
                            if (!Scopes.isCircularProxy(t)) {
                                context.map.put(key, t != null ? t : NullableObject.INSTANCE);
                            }
                        }
                        return t;
                    }
                }

                String name = key.toString();
                synchronized (currentContext) {
                    Object obj = currentContext.args.get(name);
                    if (obj == NullableObject.INSTANCE) {
                        return null;
                    }
                    T t = (T) obj;
                    if (t == null) {
                        t = provider.get();
                        if (!Scopes.isCircularProxy(t)) {
                            currentContext.args.put(name, t != null ? t : NullableObject.INSTANCE);
                        }
                    }
                    return t;
                }
            }
        };
    }

    @Override
    public String toString() {
        return "Http.Context.ARGS";
    }

    private static class Context implements ContextScoper {
        final Map<Key, Object> map = Maps.newHashMap();
        final Lock lock = new ReentrantLock();

        @Override
        public CloseableScope open() {
            lock.lock();
            final Context previous = httpContextScopeContext.get();
            httpContextScopeContext.set(this);
            return new CloseableScope() {
                @Override
                public void close() {
                    httpContextScopeContext.set(previous);
                    lock.unlock();
                }
            };
        }
    }
}
```

followed by 

```
public class AppModule extends AbstractModule {
    @Override
    protected void configure() {
        HttpContextScope httpContextScope = new HttpContextScope();
        bindScope(HttpContextScoped.class, httpContextScope);
    }

    @Provides
    @HttpContextScoped
    public TheThing providesTheThing() {
        return new TheThing();
    }
}
```

from http://stackoverflow.com/a/34981902/5266 -- so if there's a way to hide the Guiceyness of it or provide it as a custom module, that could work.

**Comment by gmethvin** [comment]
@wsargent I'm not sure how this solves the issue though. If you're using a ThreadLocal, it still makes it difficult to know whether the context will be available in a particular place. If we force it to be passed explicitly, there is no ambiguity.

**Comment by wsargent** [comment]
Yeah -- adding a Guice scope means using another ThreadLocal under the hood, so that's a problem.  (I wish Java had implicits.)

**Comment by hepin1989** [comment]
I think a better way is to deprecate the Http.context.current in 2.6,and introduce the `context -> Future[Response]` one。
if I remember correct ,there is ServiceCall<Req,Rep> in lagom。

**Comment by benmccann** [comment]
I like @wsargent's approach above of making `Context` an optional feature

**Comment by gmethvin** [comment]
I have a proof of concept here: https://github.com/gmethvin/play-java-functional-actions

We need to add some new API to make this work well, like a `withSession`/`withFlash` on Java `Result`, and new Java API for action helpers like `Cached`.

**Comment by benmccann** [comment]
Having to wrap everything in an `action()` call is more verbose than the API we have today and also will make it difficult to migrate. I suppose it'd be fine if it were optional and you wanted to muck with the `BodyParser` or something, but I'm not sure I'd want to do it for all my controllers.

**Comment by mkurz** [comment]
I agree with @benmccann: I too like @wsargent's idea of simply making `Context` an optional param. It's easy to migrate. Plus it isn't as verbose like when wrapping everything in an `action()`.

Anyway, I will provide two pull requests anytime soon which will get rid of the need to use `Http.Context.current()` in errorhandlers or `Validators`.

**Comment by mkurz** [comment]
As one part of the task of removing the thread local I opened #8306 because when the thread local is gone we also have to somehow still have access to the context from within Java error handlers.

**Comment by mkurz** [comment]
Just referencing a discussion we just had on the contributors channel starting here:
https://gitter.im/playframework/contributors?at=5ab3fb3bfa066c53255dfcf1

**Comment by mkurz** [comment]
@richdougherty @schmitch @gmethvin @marcospereira 
I am continuing [the discussing](https://gitter.im/playframework/contributors?at=5ab3fb3bfa066c53255dfcf1) we had on the gitter channel 2 days ago about the removal of the `Http.Context` and the routing problem it brings here:
What do you think about a complete different approach - like a user has to manually mark a route with a modifier  (let's for now make it a `+`) to indicate he/she wants to get the `request` object passed to his/her action method (let's say always as first parameter):
```
GET    /            +controllers.SomeController.index()
```
and now the request will be passed to the action `Result index(Request req)`.
Now the user also has the responsibility that this action method with the request param exists, otherwise, like before, compilation just fails.
So there is no need to check the action methods during routes generation via reflection or whatsoever, we have all the information we need from the parsed routes file.

So it's all the user's responsibility, Play doesn't need to automagically figure out if there is a request param and if so pass the request etc.
Right now I kind of like the idea, the users says "Pass me the request for that route, so I mark it with `+` and I also add the Request in my action, done".

What do you guys think?

**Comment by mkurz** [comment]
That would basically be a feature like we already had before in the `routes` file with the `@` prefix:
https://www.playframework.com/documentation/2.6.x/Migration25#Replaced-static-controllers-with-dependency-injection

The advantage is that we don't need to check anything, just generate the routes file and let the compiler do the work and check if the action methods a there.

**Comment by schmitch** [comment]
What if a User writes `+` in front of a Java Route?
What if `Request` is not the first param? So if I have `+controllers.SomeController.index(id: Long)`
it would get compiled to `controllers.SomeController.index(req, id)`
I'm not sure if that solution is that good. We would still rely on the Scala compiler and in error cases it would be even more cryptic than before.

Well it would probably faster to implement.
We could than generated something like:
`controllers_Assets_versioned3_invoker.call(req => Assets_0.versioned(req, path, file))`
if a `+` is in front.

and then we need to copy:
`implicit def wrapJava: HandlerInvokerFactory[JResult]` and `implicit def wrapJavaPromise: HandlerInvokerFactory[CompletionStage[JResult]]`

and create two new versions like:
`implicit def wrapJava: HandlerInvokerFactory[JRequest => JResult]` and
`implicit def wrapJava: HandlerInvokerFactory[JRequest => CompletionStage[JResult]]`.

Well even if that would be a fast solution I would still try to make a demo to see if the "reflection" based Route Generator might be another good idea.

**Comment by mkurz** [comment]
> What if a User writes `+` in front of a Java Route?

That is what I was talking about, Java routes. You mean Scala routes? Just ignore it, it's a Java routes feature only. And I think we know and can distinguish when we generate the code for a Java route or for a Scala route. Or not? If not I mean that would probably fail during compilation anyway, it's just meant for Java.

> What if Request is not the first param? So if I have `+controllers.SomeController.index(id: Long)`
it would get compiled to controllers.SomeController.index(req, id)

Exactly. It would be a convention. `+` would always mean give me the request as first parameter.
So
```
GET    /    +controllers.SomeController.index(id: Long, ...)
```
would just be a shortcut for
```
GET    /    +controllers.SomeController.index(request: Request, id: Long, ...)
```
with the only difference the the `Request` object would not come from the url params but provided by Play.

**Comment by mkurz** [comment]
However that is just and idea I had and wanted to share.

**Comment by schmitch** [comment]
You mean Scala routes? Just ignore it, it's a Java routes feature only. And I think we know and can distinguish when we generate the code for a Java route or for a Scala route. Or not?
1. Yes, I meant Scala
2. No, we can't (with the reflective Router we might could, if it's even possible)

Actually I think your idea is a good alternative.
I think to actually we solve this problem we should try out more than one solution and decide which works best. (because it's such a radical change).

**Comment by mkurz** [comment]
Alright, I gave it a try and I think #8312 looks promising, I think we can eventually just remove `Http.Context` entirely (because it basically is just an enhanced request, and all those "enhancements" can be achieved in another way). There will be a small migration task, but not too bad.

**Comment by mkurz** [comment]
Fixed in #8639 (check the migration guide I added in that pull request)
